base/src/main/java/io/spine/type/KnownTypes.java

```diff
@@ -225,6 +239,19 @@ private Type get(TypeName name) throws UnknownTypeException {
         return result;
     }
 
+    /**
+     * Gathers package names of all known message types.
+     */
+    private static ImmutableSet<PackageName> packageNames(TypeSet typeSet) {
```

Comment:
Have a look at `io.spine.reflect.PackageGraph`. We may have something similar for proto packages, and then we don't have to create sub-packages for each of the package. We'd have the full picture created once.